src/datachain/lib/dc.py

```diff
@@ -331,22 +346,80 @@ def from_storage(
             chain = DataChain.from_storage("s3://my-bucket/my-dir")
             ```
         """
-        func = get_file(type)
-        return (
-            cls(
-                path,
+        file_type = get_file_type(type)
+
+        if client_config is None:
+            client_config = {}
+
+        if anon:
+            client_config["anon"] = True
+
+        session = Session.get(
+            session,
+            catalog=kwargs.get("catalog"),
+            client_config=client_config,
+            in_memory=in_memory,
+        )
+
+        client, path = Client.parse_url(
+            uri, session.catalog.cache, **session.catalog.client_config
+        )
+
+        # clean path without globs
+        lst_path = (
+            posixpath.dirname(path)
+            if glob.has_magic(os.path.basename(os.path.normpath(path)))
+            or client.fs.isfile(uri)
+            else path
+        )
+        lst_uri = f"{client.uri}/{lst_path.lstrip('/')}"
+
+        ds_name = listing_dataset_name(client.uri, posixpath.join(lst_path, ""))
+
+        for ds in cls.datasets(
+            session=session, in_memory=in_memory, include_listing=True
+        ).collect("dataset"):
+            if listing_expired(ds.created_at):  # type: ignore[union-attr]
+                continue
+            if (
+                listing_subset(ds.name, ds_name)  # type: ignore[union-attr]
+                and not update
+            ):
+                # we can use found listing as it contains the one from input
+                dc = cls.from_dataset(ds.name, **kwargs)  # type: ignore[union-attr]
+                dc.signals_schema = dc.signals_schema.mutate(
+                    {f"{object_name}": file_type}
+                )
+                return ls(dc, path, recursive=recursive)
+
+        # caching new listing to special listing dataset
+        (
+            cls.from_records(
+                DataChain.DEFAULT_FILE_RECORD,
                 session=session,
-                recursive=recursive,
-                update=update,
                 in_memory=in_memory,
                 **kwargs,
             )
-            .map(**{object_name: func})
-            .select(object_name)
+            .gen(
+                list_bucket(lst_uri, **session.catalog.client_config),
+                output={f"{object_name}": File},
+            )
+            .save(ds_name, listing=True)
         )
 
+        dc = cls.from_dataset(ds_name, session=session, **kwargs)
+        dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
+
+        return ls(dc, path, recursive=recursive, object_name=object_name)
+
     @classmethod
-    def from_dataset(cls, name: str, version: Optional[int] = None) -> "DataChain":
+    def from_dataset(
+        cls,
+        name: str,
+        version: Optional[int] = None,
+        session: Optional[Session] = None,
```

Comment:
why do we need session now here?

Reply:
yes, we need session and we don't need kwargs. See my other comments on Session, Catalog and client_config. In short - we can afford only one config object in user API, not 3.

Reply:
@ilongin same question here - can we iterate / discuss the API?

Reply:
I put my thoughts in the comment below https://github.com/iterative/datachain/pull/294#discussion_r1730464596